### ISSUE TITLE
Have tools work better with out-of-spec GTFs (often fall back to the GFF datatype). Note: GFF != GFF3

### ISSUE THREAD

**Issue by jennaj** [galaxyproject/galaxy#7947]
cc @jmchilton @nsoranzo 

**Underlying issue**: 

* Headers in `GTF` files are common in the public domain but are out of file specification.
* `GFF` is the fall back datatype assigned by Upload > Type > "Autodetect" and Edit Attributes > Datatypes > "Detect Datatype" whenever a GFF2/GTF/GFF3 file that is out-of-spect is input.

**Potential solution**:

**We could make a converter for `GFF` > `GTF` and prevent *many* tool errors**.
* If the problematic `GFF/GTF` headers are removed, the files will auto/redetect as `GFF` (only meet that specification) or `GTF` (if meet the extra specification requirements). 
* Standalone converter tool + Edit Attributes convert function + an implicit converter for tools that accept `GTF`.
* If the file does not meet `GTF` specification (after removing headers), the conversion fails at that point. 
* If the file does not meet `GFF3` specification, that is not what this converter is doing, is a different problem, and more complex to solve (see linked ticket below). I vote to skip that for now, or at least have it be a distinct converter tool that is named correctly (GFF3 to GTF).
* This should be exposed to the user (unhide the converted dataset, pass or fail). They should know what is going on.
* Not sure if we could prevent a tool from still running and also erroring due to unusable inputs (intermediate converter step failed), but maybe doesn't matter _IF the failed conversion is not hidden in the history panel_ and has some kind of meaningful error message. Users do not know to look for hidden implicitly converted datasets when something goes wrong.

**Some related tickets:** (are plenty more)

Why so many tickets? `GFF` is being mixed up with `GFF3` -- at least in part. Some tools _accept_ any of `GFF/GTF/GFF3` and some _actually work_ with any and some _only work with certain types that are actually in spec for the assigned datatype_.

1. This looks Ok and there is probably not more we can or should do at the sniffer level:
https://github.com/galaxyproject/galaxy/pull/3832

2. This one mixed up `GFF/GFF3` so was really for a GFF3 > GTF converter. I think we could use both but as a secondary, distinct, and probably lower priority. Accepting `GFF` as a general catch-all when `GTF` is required by the tool is what is triggering the rafts of tool errors. Users will change the `GFF` datatype themselves to `GTF` to try to get it right (have tool recognize the dataset) only to have jobs fail out, usually because of the headers. :|
https://github.com/galaxyproject/galaxy/pull/430

**Let's get these GTF-with-header data to work.** 

Public sources for `GTF` data should "just work" in Galaxy. They are everywhere. If just a teeny bit off-spec in an obvious way, we can smooth that out. The public sources are including headers to add in details about what the data represents (source/content-wise). I don't see that changing any time in the future and we don't want to just strip it out ourselves at Upload (information loss). We should instead work with the data "as-is" in a smart way, and without re-writing every tool wrapper that errors out for this specific reason. 

Even Wikipedia is a bit confused: https://en.wikipedia.org/wiki/General_feature_format

Yes, I didn't mention GFF2 on purpose. Is deprecated, replaced by GTF. Handling GFF/GTF/GFF3 correctly is enough to solve the usage issues impacting Galaxy users that I have seen.

**How the wider Bioinfo community got here :sigh:**

![IMG_0347](https://user-images.githubusercontent.com/8999735/57648721-124d6500-757b-11e9-9ff4-6d9bff7ab28c.jpeg)

**Comment by jennaj** [comment]
Related big picture solution for data validation in general https://github.com/galaxyproject/galaxy/pull/7913

**Comment by nsoranzo** [comment]
@jennaj I've looked at the [GTF sniffer code](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/datatypes/interval.py#L1004) and tested it, I'm pretty sure that headers don't prevent a file from being sniffed as GTF. Can you please share an example file that fails to be recognised as GTF when uploaded?

**Comment by jennaj** [comment]
Dataset #1 in this test history https://usegalaxy.org:/u/jen/h/test-gft-sniffed-as-gff is a GTF, with a header, and it is sniffed as GFF. 

Removing the header is not enough to detect this data as GTF, but that's probably because of other problems (the "gene" lines only have gene_id, no transcript_id). BUT -- if GTF is directly assigned, _with or without the header removed_ && _even with the problematic "no transcript_id" lines_ -- the data works with tools. **This is new and possibly related to 19.05** changes (we put 19.05 with bits of 19.09 on main yesterday).

GencodeGenes is a good source for human/mouse data (not as hard to get uploaded to Galaxy as iGenomes).

**This changes the request.** The GFF > GTF converter could only keep lines that contain both gene_id and transcript_id. The correct datatype can then the autodetected by Edit Attributes (or directly assigned by the converter). I don't think we should change Upload/sniffers. I don't think that removing headers will cause harm (tools other than what I just tested might still not like those). 

This is a cleaner approach with fewer assumptions about what any particular data claims to be from the source. I also don't think we should just pull out "transcript" or "exon" lines, tools that use this data can filter around it (as well as "gene" lines). 

- An example of filtering for "transcript_id" (only) is in data 62-63-64 in the same test history. It autodetects as GTF after.

- An example of filtering out problematic "gene" lines (don't have transcript_id) but leaving in headers is in data 68. It autodetects as GTF after -- even though it has headers. _This is new!_ "filtering out problematic "gene" lines (don't have transcript_id)" is the same thing for this data as filtering FOR transcript_id + gene_id -- but I think the "filtering FOR transcript_id + gene_id" is better for the converter.

**Some tools do only accept data that is assigned GTF for select fields (and not GFF, for whatever reason it was rejected as GTF), so still think the converter is worth it.**

More details because the header thing threw me off. I couldn't replicate the prior errors! (a good thing)

**I don't think that headers matter anymore as far as tool errors go**. Tried many data/format combinations and the tools are dealing with the headers fine now (yea!!). 

In short: 

Header is OK with all now of these tools (19.05 changes?)
- **HISAT2 GTF (not GFF/GFF3)**
- HTSeq GFF/GTF (not GFF3)
- Featurecounts GFF/GTF/GFF3
- **Star GTF/GFF3 (not GFF)** 
- **Stringtie GTF/GFF3 (not GFF)** 

For HISAT2/STAR/Stringtie, using an implicit GFF>GTF converter for files that confirmed to have gene_id/transcript_id only lines (and filtered to only preserve those) could be used without the user needing to manipulate data from data sources that produce slightly out-of-spec GTFs, without breaking what GTF means to Galaxy's sniffers, and without triggering tool errors.

**Comment by jmchilton** [comment]
> Dataset #1 in this test history https://usegalaxy.org:/u/jen/h/test-gft-sniffed-as-gff is a GTF, with a header, and it is sniffed as GFF.

It isn't a GTF in the Galaxy sense - we use this definition (http://genome.ucsc.edu/FAQ/FAQformat#format4) and that file simply doesn't match that format - it doesn't have the required attributes.

**Comment by jennaj** [comment]
Agree.

Given that the most commonly used tools now can handle the GFF/GTF headers, and work with some slightly out-of-spec GTFs (with a manually assigned datatype), maybe can skip this converter for now. Headers were the primary (and prolific!) problem before, and the initial reason for this request. 

Going to close the ticket out. We can reopen if needed later for some reason. I'll also adjust our FAQ help to match current behavior.
